src/System.Text.Json/tests/Serialization/CustomConverterTests.DerivedTypes.cs

```diff
@@ -0,0 +1,187 @@
+            {
+                Assert.Equal(expected++, value);
+            }
+            Assert.Equal(1, wrapper.DictionaryWrapper["key"]);
+            Assert.Equal(2, wrapper.Dictionary["key"]);
+
+            string serialized = JsonSerializer.Serialize(wrapper);
+            Assert.Contains(@"""ListWrapper"":[1,2,3]", serialized);
+            Assert.Contains(@"""List"":[4,5,6]", serialized);
+            Assert.Contains(@"""DictionaryWrapper"":{""key"":1}", serialized);
+            Assert.Contains(@"""Dictionary"":{""key"":2}", serialized);
+
+            // With converters, we expect no values in the wrappers per converters' implementation.
+            JsonSerializerOptions options = new JsonSerializerOptions();
+            options.Converters.Add(new ListWrapperConverter());
+            options.Converters.Add(new DictionaryWrapperConverter());
+
+            DerivedTypesWrapper customWrapper = JsonSerializer.Deserialize<DerivedTypesWrapper>(json, options);
+            Assert.Null(customWrapper.ListWrapper);
+            expected = 4;
+            foreach (int value in customWrapper.List)
+            {
+                Assert.Equal(expected++, value);
+            }
+            Assert.Null(customWrapper.DictionaryWrapper);
+            Assert.Equal(2, customWrapper.Dictionary["key"]);
+
+            serialized = JsonSerializer.Serialize(wrapper, options);
+            Assert.Contains(@"""ListWrapper"":[]", serialized);
+            Assert.Contains(@"""List"":[4,5,6]", serialized);
+            Assert.Contains(@"""DictionaryWrapper"":{}", serialized);
+            Assert.Contains(@"""Dictionary"":{""key"":2}", serialized);
+        }
+
+        [Fact]
+        public static void CustomUnsupportedDerivedTypeConverter()
+        {
+            string json = @"{""DictionaryWrapper"": {""1"": 1}}";
+
+            UnsupportedDerivedTypesWrapper wrapper = new UnsupportedDerivedTypesWrapper
+            {
+                DictionaryWrapper = new UnsupportedDictionaryWrapper()
+            };
+            wrapper.DictionaryWrapper[1] = 1;
+
+            // Without converter, we throw.
+            Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<UnsupportedDerivedTypesWrapper>(json));
+            Assert.Throws<NotSupportedException>(() => JsonSerializer.Serialize(wrapper));
+
+            // With converter, we expect no values in the wrapper per converter's implementation.
+            JsonSerializerOptions options = new JsonSerializerOptions();
+            options.Converters.Add(new UnsupportedDictionaryWrapperConverter());
+
+            UnsupportedDerivedTypesWrapper customWrapper = JsonSerializer.Deserialize<UnsupportedDerivedTypesWrapper>(json, options);
+            Assert.Null(customWrapper.DictionaryWrapper);
+
+            Assert.Contains(@"""DictionaryWrapper"":{}", JsonSerializer.Serialize(wrapper, options));
+        }
+    }
+
+    public class ListWrapper : List<int> { }
```

Comment:
Add tests for IEnumerable as well (from https://github.com/dotnet/corefx/issues/40391).